### PR TITLE
Updated to v1.14.3

### PR DESCRIPTION
**Pull request by anmol26s**
## Problem

- *New update available*

## Solution

- *Updated to v1.14.3*

### CODE REVIEW

**Comment by anmol26s** [comment]
!testme

**Comment by anmol26s** [comment]
!testme

**Comment by anmol26s** [comment]
!testme

**Comment by kevinpapst** [comment]
There is a Ruby script which works: https://www.kimai.org/blog/2021/sqlite-and-ftp-support-removed/ 
But I would rather disallow migration from SQLite and show an explanation - doing the migration automatically would be great but feels a bit risky.

**Comment by anmol26s** [comment]
That would be lot of work and extra effort to migrate from Sql to Mysql. I don't know if its worth to make that much changes.
So,
1. We can check if the database is Sqlite. We can stop the upgrade process with an error message that Sqlite is not supported anymore. This is like saying that game ends here, start over again. I don't think there would many users on Sqlite, when you have option for mysql from day 1.
2. Or we can try to force migrate everyone to mysql with above method.

We can have a poll on Yuno to see if there are potential users that needs to migrated to Mysql.

**Comment by yalh76** [comment]
> 
> > Why doing a partial switch of source to git ? Doing that we don't know which Kimai2 version will be installed ...
> 
> It's only for new installations. I don't know if its good idea to bring everyone to git.If you recommend we can try to move everyone to git. I did not do it because I don't have live running Kimai2 app. So there is no way to test that this would not break anything in while we migrate to git. We can always check the git tag to test the version of installed app.
> We can give option to use git or not to use it, while installation if installing with git is not OK with you.

If you think that moving everyone to git is not a good idea, better to let all with app.src. But having to maintain two different source system will be difficult
The point is there two options for me:
1. using app.src: 
  - Pro: each new installation install a precise version of the upstream app
  - Cons: if upstream app provide a new version, package must be updated
2. using git
  - Pro: package is always up to date with the upstream app
  - Cons: each new installation can install an unknown new version, with new unmet requirements... Install may fail...

**Comment by kevinpapst** [comment]
As being "the upstream" 😁 I can tell that the only requirements that changed in the last 1,5 years was:
- the drop of SQLITE (which was always announced as being for test only, not production - my mistake that I didn't see it here earlier) 
- the change of minimum PHP version (being 7.2.9 now and the next version bumps it to PHP 7.3). So rather conservative as 7.2 support ended half a year ago and I guess YunoHost is more up-to-date?

**Comment by yalh76** [comment]
> 
> 
> As being "the upstream" 😁 I can tell that the only requirements that changed in the last 1,5 years was:
> 
>     * the drop of SQLITE (which was always announced as being for test only, not production - my mistake that I didn't see it here earlier)
> 
>     * the change of minimum PHP version (being 7.2.9 now and the next version bumps it to PHP 7.3). So rather conservative as 7.2 support ended half a year ago and I guess YunoHost is more up-to-date?

It was a more general comment, not relative to Kimai ^^ because @anmol26s made several PR on several apps to switch them to git...

And the question is:
Do we prefer to have some packages to be late from the upstream app or do we prefer some of them having installation fail suddenly...

**Comment by kevinpapst** [comment]
Ok, I silence myself 😁 this sounds like a general question that needs to be answered somewhere else.

**Comment by yalh76** [comment]
> 
> 
> Ok, I silence myself 😁 this sounds like a general question that needs to be answered somewhere else.

Sorry, it was in Kimai upgrade but I made a more general statement ^^

**Comment by anmol26s** [comment]
Git: Its more easy to change ecosystems.

> Cons: each new installation can install an unknown new version, with new unmet requirements... Install may fail...

We can still use tags to force the package to stay on a particular version.

**Comment by yalh76** [comment]
> 
> 
> Git: Its more easy to change ecosystems.
> 
> > Cons: each new installation can install an unknown new version, with new unmet requirements... Install may fail...
> 
> We can still use tags to force the package to stay on a particular version.

Good point using tags to specify the version to install and having previous app.src install moving to git source + manage the end of sqlite ^^

**Comment by anmol26s** [comment]
!testme

**Comment by anmol26s** [comment]
Do we need to migrate the sqlite to mysql? I think its better to give an error message and stop the upgrade if the instance is on sqlite.

File: manifest.json [hunk 1]
```diff
@@ -68,6 +68,15 @@
...
+            },
+            {
+                
+           "name": "update",
+           "type": "boolean",
+           "ask":{
+                "en": "Making this option to true will give you an option to force upgrade to new version without the package being updated to that version. WARNING: This can be risky and break your app if specific version have additional steps than noraml upgrade. Keep the option false, if you are not sure."
+           },
+           "default": false
             }
```
Comment (yalh76) [inline manifest.json:80]: I don't think it's a good idea to let people install a newer version than the last tested... Or we will have the same issue as letting all people going to an unknown version

**Review by yalh76 — CHANGES_REQUESTED** [review]

File: manifest.json (hunk 1 above)
Comment (anmol26s) [reply manifest.json:80]: Ok

**Comment by anmol26s** [comment]
Ok, no more option to force update.
@yalh76

**Comment by anmol26s** [comment]
!testme

File: scripts/_common.sh [hunk 2]
```diff
@@ -4,6 +4,9 @@
 # COMMON VARIABLES
 #=================================================
 
+# Latest tag
+latest_tag=$(curl -s https://api.github.com/repos/kevinpapst/kimai2/releases/latest | grep 'tag_name' | cut -d\" -f4)
+tag="1.14.3"
```
Comment (yalh76) [inline scripts/_common.sh:9]: It should be moire secure to use the commit ID corresponding of the TAG instead of the TAG

File: scripts/install [hunk 3]
```diff
@@ -29,10 +29,9 @@ path_url="/"
 admin=$YNH_APP_ARG_ADMIN
 is_public=$YNH_APP_ARG_IS_PUBLIC
 registration=$YNH_APP_ARG_REGISTRATION
-database=$YNH_APP_ARG_DATABASE
 random_key=$(ynh_string_random 32)
-
 app=$YNH_APP_INSTANCE_NAME
+update=0
```
Comment (yalh76) [inline scripts/install:34]: no more needed

File: scripts/install [hunk 4]
```diff
@@ -73,8 +72,10 @@ ynh_install_app_dependencies $pkg_dependencies
 ynh_script_progression --message="Setting up source files..."
 
 ynh_app_setting_set --app=$app --key=final_path --value=$final_path
-# Download, check integrity, uncompress and patch the source from app.src
-ynh_setup_source --dest_dir="$final_path"
+
+# Clone the latest version
+
+git clone -b $tag --depth 1 https://github.com/kevinpapst/kimai2.git $final_path
```
Comment (yalh76) [inline scripts/install:78]: More secure to use the commit ID

File: scripts/install [hunk 5]
```diff
@@ -154,7 +149,7 @@ popd
 #=================================================
 # STORE THE CONFIG FILE CHECKSUM
 #=================================================
-ynh_script_progression --message="Storing the config file checksum..."
+ynh_script_progression  --message="Storing the config file checksum..."
```
Comment (yalh76) [inline scripts/install:152]:
```suggestion
ynh_script_progression --message="Storing the config file checksum..."
```

File: scripts/install [hunk 6]
```diff
@@ -175,9 +170,9 @@ chown -R $app: $final_path
 #=================================================
 ynh_script_progression --message="Configuring SSOwat..."
 
-ynh_permission_create --permission="teamlead"
-ynh_permission_create --permission="admin"
-ynh_permission_create --permission="super_admin" --allowed "$admin"
+ynh_permission_create  --permission="teamlead"
+ynh_permission_create  --permission="admin"
+ynh_permission_create  --permission="super_admin" --allowed "$admin"
```
Comment (yalh76) [inline scripts/install:175]:
```suggestion
ynh_permission_create --permission="teamlead"
ynh_permission_create --permission="admin"
ynh_permission_create --permission="super_admin" --allowed "$admin"
```

File: scripts/install [hunk 7]
```diff
@@ -110,42 +111,36 @@ ynh_script_progression --message="Configuring Kimai2..."
...
 
 #=================================================
 # BUILD KIMAI2
 #=================================================
 ynh_script_progression --message="Building Kimai2..."
 
-ynh_install_composer --phpversion="${YNH_PHP_VERSION}" --workdir="$final_path" --install_args="--optimize-autoloader"
-ynh_composer_exec --phpversion="${YNH_PHP_VERSION}" --workdir="$final_path" --commands="require laminas/laminas-ldap"
+ynh_install_composer   --phpversion="${YNH_PHP_VERSION}" --workdir="$final_path" --install_args="--optimize-autoloader"
+ynh_composer_exec      --phpversion="${YNH_PHP_VERSION}" --workdir="$final_path" --commands="require laminas/laminas-ldap"
```
Comment (yalh76) [inline scripts/install:143]:
```suggestion
ynh_install_composer --phpversion="${YNH_PHP_VERSION}" --workdir="$final_path" --install_args="--optimize-autoloader"
ynh_composer_exec --phpversion="${YNH_PHP_VERSION}" --workdir="$final_path" --commands="require laminas/laminas-ldap"
```

File: scripts/install [hunk 8]
```diff
@@ -110,42 +111,36 @@ ynh_script_progression --message="Configuring Kimai2..."
...
+cp ../conf/.env.mysql $env_conf
+db_name=$(ynh_sanitize_dbid --db_name=$app)
+db_user=$db_name
+ynh_app_setting_set --app=$app                       --key=db_name --value=$db_name
+ynh_mysql_setup_db  --db_user=$db_user               --db_name=$db_name	
+ynh_replace_string  --match_string="__RANDOM_KEY__"  --replace_string="$random_key" --target_file="$env_conf"
+ynh_replace_string  --match_string="__DOMAIN__"      --replace_string="$domain"     --target_file="$env_conf"
+ynh_replace_string  --match_string="__DB_PASSWORD__" --replace_string="$db_pwd"     --target_file="$env_conf"
+ynh_replace_string  --match_string="__DB_USER__"     --replace_string="$db_name"    --target_file="$env_conf"
+ynh_replace_string  --match_string="__DB_NAME__"     --replace_string="$db_name"    --target_file="$env_conf"
```
Comment (yalh76) [inline scripts/install:135]: Would be better to use ynh_add_config

File: scripts/remove [hunk 9]
```diff
@@ -18,21 +18,21 @@ app=$YNH_APP_INSTANCE_NAME
 
 domain=$(ynh_app_setting_get --app=$app --key=domain)
 final_path=$(ynh_app_setting_get --app=$app --key=final_path)
-database=$(ynh_app_setting_get --app=$app --key=database)
+
```
Comment (yalh76) [inline scripts/remove:21]:
```suggestion
```

File: scripts/upgrade [hunk 10]
```diff
@@ -72,9 +72,21 @@ ynh_abort_if_errors
...
+        if [ $update -eq 1 ]
+        then
+          git fetch origin tag $latest_tag --no-tags
+          git checkout $latest_tag  
+        else
+          git fetch origin tag $tag --no-tags
+          git checkout $tag
+        fi
+      popd
+  else
```
Comment (yalh76) [inline scripts/upgrade:89]: Should be rewrite to only use GIT

File: scripts/upgrade [hunk 11]
```diff
@@ -161,23 +183,20 @@ ynh_store_file_checksum --file="$local_conf"
...
+
+db_pwd=$(ynh_app_setting_get --app=$app --key=mysqlpwd)
+db_name=$(ynh_app_setting_get --app=$app --key=db_name)
+db_user=$db_name	
+ynh_replace_string --match_string="__RANDOM_KEY__" --replace_string="$random_key" --target_file="$env_conf"
+ynh_replace_string --match_string="__DOMAIN__" --replace_string="$domain" --target_file="$env_conf"
+ynh_replace_string --match_string="__DB_PASSWORD__" --replace_string="$db_pwd" --target_file="$env_conf"
+ynh_replace_string --match_string="__DB_USER__" --replace_string="$db_name" --target_file="$env_conf"
+ynh_replace_string --match_string="__DB_NAME__" --replace_string="$db_name" --target_file="$env_conf"
+
```
Comment (yalh76) [inline scripts/upgrade:199]: Would be better to use ynh_add_config

**Review by yalh76 — CHANGES_REQUESTED** [review]

**Comment by anmol26s** [comment]
!testme

**Review by ericgaspar — APPROVED** [review]

**Pull request merged**